flake.nix

```diff
@@ -13,56 +13,112 @@
-    };
-
-    haskell-flake-utils = {
-      url = "github:ivanovs-4/haskell-flake-utils";
-      inputs.flake-utils.follows = "flake-utils";
-    };
   };
 
-outputs = { self, nixpkgs, flake-utils, haskell-flake-utils, flake-compat, ... }:
-  haskell-flake-utils.lib.simpleCabalProject2flake {
-    inherit self nixpkgs;
+  outputs = inputs:
+    with builtins;
+    let
+      lib = inputs.nixpkgs.lib;
+      pnames = [
+        "automaton"
+        "rhine"
+        "rhine-bayes"
+        "rhine-examples"
+        "rhine-gloss"
+        "rhine-terminal"
+      ];
+      supportedGhcs = [ "ghc92" "ghc94" "ghc96" "ghc98" ];
+      libPnames = filter (pname: pname != "rhine-examples") pnames;
+      hpsFor = pkgs:
+        lib.filterAttrs (name: _: elem name supportedGhcs) pkgs.haskell.packages
+        // { default = pkgs.haskellPackages; };
+      rhinePackages = hfinal: hprev:
+        lib.genAttrs pnames (pname: hfinal.callCabal2nix pname ./${pname} { });
+      overlay = final: prev:
+        let hps = hpsFor final; in {
+          haskell = prev.haskell // {
+            packageOverrides = with prev.haskell.lib.compose; lib.composeManyExtensions [
+              prev.haskell.packageOverrides
+              rhinePackages
+              (hfinal: hprev: {
+                monad-bayes = markUnbroken hprev.monad-bayes;
+                monad-schedule = hprev.callHackageDirect
+                  {
+                    pkg = "monad-schedule";
+                    ver = "0.2";
+                    sha256 = "sha256-Z9lAxkvJDH9aQZd65bGOQI3EGH7oSAhK0nuBKULgiCE=";
+                  }
+                  { };
+                time-domain = hprev.callHackageDirect
+                  {
+                    pkg = "time-domain";
+                    ver = "0.1.0.5";
+                    sha256 = "sha256-llDBQuU5ez/0MiOIMH97P4BQhFDyPfTMWinq1wJrDGI=";
+                  }
+                  { };
+              })
+              (hfinal: hprev: lib.optionalAttrs prev.stdenv.isDarwin {
+                monad-schedule = dontCheck hprev.monad-schedule;
+              })
+              (hfinal: hprev: lib.optionalAttrs (lib.versionOlder hprev.ghc.version "9.4") {
+                time-domain = doJailbreak hprev.time-domain;
+              })
+            ];
+          };
+          rhine-bin = prev.buildEnv {
+            name = "rhine-bin";
+            paths = map (pname: hps.default.${pname}) pnames;
+            pathsToLink = [ "/bin" ];
+          };
+          rhine-lib = prev.buildEnv {
+            name = "rhine-lib";
+            paths = lib.mapCartesianProduct
+              ({ hp, pname }: hp.${pname})
+              { hp = attrValues hps; pname = pnames; };
+            pathsToLink = [ "/lib" ];
+          };
+          rhine-docs = prev.buildEnv {
+            name = "rhine-docs";
+            paths = map (pname: prev.haskell.lib.documentationTarball hps.default.${pname}) libPnames;
+          };
+          rhine-sdist = prev.buildEnv {
```

Comment:
Can I access these `rhine-*` derivations directly in any way? Or are they just prerequisites for `rhine-all`?

Reply:
Good point. I've exposed them by simply making `legacyPackages` point to `pkgs`. This way we can access everything in our overlay.

Reply:
Sorry, I had to force-push over that change. Can you apply it again?